### PR TITLE
Redo partition loops

### PR DESCRIPTION
**Pull request by abadams**
This PR partitions at every level from the outermost in, and also
partitions GPU loops. 

The previous loop partitioning method (only partition innermost loop)
fell on its face for things like convolutions - the loop over the kernel
was split up instead of the loop over the image, so some of our code 
is still littered with the explicit scheduling of padded versions of the input.

The code expansion factor is linear in the number of dimensions you put
a boundary condition on.

Included a benchmark that tests a tiled convolution with a runtime size.
If you never hit the boundaries, the boundary conditions are all free.
They cost a moderate amount if you do hit the boundaries a bunch. On the
GPU they seem to be free all the time - it generates a kernel with a branch
on a block variable.

### CODE REVIEW

File: src/DeviceInterface.cpp [hunk 1]
```diff
@@ -80,7 +80,13 @@ int halide_copy_to_device(void *user_context, struct buffer_t *buf,
 /** Wait for current GPU operations to complete. Calling this explicitly
  * should rarely be necessary, except maybe for profiling. */
 int halide_device_sync(void *user_context, struct buffer_t *buf) {
-    return false;
+    user_assert(user_context == NULL) << "Cannot provide user_context to libHalide.a halide_device_sync\n";
```
Comment (steven-johnson) [inline src/DeviceInterface.cpp]: Why is user_context forbidden here?

Comment (abadams) [inline src/DeviceInterface.cpp]: The functions in DeviceInterface.cpp are the versions that gets called when
jitting. In that case the user_context is hijacked for other purposes. At
least I think that's what's going on - I copied-pasted this assert from the
other functions in the file.

Comment (zvookin) [inline src/DeviceInterface.cpp]: For JIT, if the user_context is non-null, it is a structure that holds the function pointers that are the callbacks for the JIT realization that made the call into the runtime. The JIT sets its own custom handlers a number of routines, such as halide_print, etc. These handlers look at the user_context and if it is non-null, indirect through it to call the desired target routine. This allows making the set_custom_* methods on Func work while sharing a single copy of the Halide runtime across JIT invocations of different Funcs.  If the user_context is nullptr, then the default runtime routines are used.

(Lower-level, the runtime code contains globals that hold the custom handers. The JIT takes the one global that exists and sets it to a routine that indirects through a table of function pointers, or calls the default routine if the pointer to the table is nullptr. The user_context mechanism is used to pass the table pointer. The table of function pointers has one pointer per type of handler.)

The wrappers in DeviceInterface.cpp are called by the client code and the client code has no way to produce the callback structure, nor should is assume details of the JIT handlers. But if the handlers are passed a non-null user_context, the JIT installed handlers would crash trying to indirect through it.

In effect, user_context is not usable for client code in JIT.

I hope we'll someday replace user_context with something cleaner and more sensible, but for the now, these are the facts of life.

File: src/PartitionLoops.cpp [hunk 2]
```diff
@@ -17,748 +19,652 @@ using std::string;
...
+        string epilogue_name = unique_name(op->name + ".epilogue", false);
 
-        if (value.same_as(op->value) && body.same_as(op->body)) {
-            expr = op;
-        } else {
-            expr = Let::make(op->name, value, body);
-        }
-    }
+        if (make_prologue) {
+            // They'll simplify better if you put them in lexicographic order
```
Comment (dsharletg) [inline src/PartitionLoops.cpp]: Can you say why this is? It's an interesting comment. Do the simplifications conflict with each other?

File: src/PartitionLoops.cpp [hunk 3]
```diff
@@ -17,748 +19,652 @@ using std::string;
...
-    Scope<Expr> bound_vars;
-    Scope<int> free_vars;
-    Scope<int> inner_loop_vars;
+bool has_likely_tag(Expr e) {
+    HasLikelyTag h;
+    e.accept(&h);
+    return h.result;
+}
 
+struct Simplification {
```
Comment (dsharletg) [inline src/PartitionLoops.cpp]: How about a comment here explaining the motivation for finding simplifications on subsets of the domain for a loop variable? I think it would help to understand what this is doing when you later understand that the steady state is taken to be the intersection of all of the subsets, where all of the simplifications can be applied.

**Comment by abadams** [comment]
This pass breaks the new fft benchmark. Will fix.

**Comment by abadams** [comment]
I have pushed a bunch of comments. Let me know if they're useful.
